### DISCUSSION TITLE
ad-hoc 401 Unauthorized errors

### DISCUSSION THREAD

**Discussion by KevinJTaylor** [lestrrat-go/jwx#551] [Q&A]
Hi all, has anyone else encountered random errors when running tests against jwx-enabled restful services? 

I'm using httpexpect to do these test, generating an access token and calling the endpoint in my test, validating the response code and if that works obviously the data itself (obviously I'm using jwx (jwk/jwt) to do the authentication/authorisation). Generally what happens is, one or two tests fail on each run due to a 401 Unauthorized error, but it is nearly always different tests. I believe the problem is down to expiry, but that is a guess, and it is difficult to be sure due to the randomness of the failures.

Any ideas?

```
    chain.go:21: 
        expected status equal to:
         "200 OK"
        
        but got:
         "401 Unauthorized"
```

**Comment by lestrrat** [comment]
I think you will need to provide more details/code. Also, do you not have control over the server side? If you do it would be much easier to sprinkle debug logs than to try to guess what the cause might be.

**Reply by KevinJTaylor** [reply]
I figured you may say that :)

I've been trying to feel my way around the issue, but as I said consistently reproducing the problem was my main blocker. This was the first thing I came across which made me wonder whether something was perhaps going on with time validation ([undefined timezone behavior in Golang](https://www.dolthub.com/blog/2021-09-03-golang-time-bugs/))

Still I had no concrete evidence, so I tried to pare back my test to see if I could reproduce the problem, and I now seem to have been able to do that. The problem is not actually with expiry but with the "issued at" validation. In effect I was getting an "iat not satisfied" error. The way I've been able to reproduce it is a little "sledge-hammer to crack a nut" but hey.
```
func TestIatIssue(t *testing.T) { 
	for i := 0; i < 1000; i++ { 
		sa, err := json.Marshal(authmod.MockServiceAccount) 
		require.Nil(t, err) 
		ts, err := idtoken.NewTokenSource(context.Background(), "me", idtoken.WithCredentialsJSON(sa)) 
		require.Nil(t, err) 
		to, err := ts.Token() 
		require.Nil(t, err) 
		at := to.AccessToken 
		_, err = jwt.ParseString( 
			at, 
			jwt.WithKeySetProvider(&ks{}), 
			jwt.InferAlgorithmFromKey(true), 
			jwt.WithValidate(true), 
		) 
		if err != nil { 
			fmt.Printf("error: %v, exp: %v", err, to.Expiry) 
			require.Nil(t, err) 
		} 
		time.Sleep(1 * time.Second) 
	} 
} 
```

What I found is that this test fails (at some point)

The first thing I did once I was able to reproduce the error was decorate the validation (this is in the isIssuedAtValid func of jwt/validate.go)
```
		if now.Before(ttv.Add(-1 * skew)) { 
			return errors.Errorf(`iat not satisfied: now=%v, ttv=%v, skew=%v`, now, ttv, skew) 
		} 
```
The output from this was "iat not satisfied: now=2022-01-15 12:02:56 +0000 GMT, ttv=2022-01-15 12:02:57 +0000 UTC, skew=0s"

You can see that the timezones in now and ttv are different, but as I'm in the good old UK that shouldn't be a problem (unless we were doing DeepEqual, should it?). The other thing you can see is that it looks like the ttv is a second after now (the skew is 0 so shouldn't affect the outcome)

Anyway, having seen this behaviour and thinking it may be related to the article above (no defined timezone), I made a simple change in the first line of the test to load the local timezone. Unfortunately, although this seems to make things a bit more robust it still didn't work ... rather than failing fairly quickly (first 10 iterations), it seems to fail later (last attempt was on iteration 47), but that may just be coincidence.

My final change was to calculate & print out the skewed ttv value on failure (sTtv) ...

iat NOT satisfied: 
now=2022-01-15 12:21:06 +0000 GMT, 
ttv=2022-01-15 12:21:07 +0000 UTC, 
sTtv=2022-01-15 12:21:07 +0000 UTC, 
skew=0s

I don't really understand why this is failing now, as it would seem on the face of it, that the calculation now.Before(sTtv) should succeed 😮‍💨 

Of course I haven't printed out the nano seconds used in the Before calculation (maybe the act of printing times in different timezones rounds differently?)

**Reply by lestrrat** [reply]
This is me thinking out lout without really checking, but it could be monotonic clock problem. We may just need to strip out the monotonic clock part. I wonder what happens if we do `t.Truncate(0)`

It's my bed time here in my timezone, so I will come back to this tomorrow onwards, just letting you know... :)

**Reply by lestrrat** [reply]
Last post from me today. What happens when you apply this?

```diff
diff --git a/jwt/validate.go b/jwt/validate.go
index c1a6894..559af3a 100644
--- a/jwt/validate.go
+++ b/jwt/validate.go
@@ -253,8 +253,8 @@ func IsIssuedAtValid() Validator {
 func isIssuedAtValid(ctx context.Context, t Token) error {
        if tv := t.IssuedAt(); !tv.IsZero() && tv.Unix() != 0 {
                clock := ValidationCtxClock(ctx) // MUST be populated
-               now := clock.Now().Truncate(time.Second)
-               ttv := tv.Truncate(time.Second)
+               now := clock.Now().Truncate(time.Second).Truncate(0)
+               ttv := tv.Truncate(time.Second).Truncate(0)
                skew := ValidationCtxSkew(ctx) // MUST be populated
                if now.Before(ttv.Add(-1 * skew)) {
                        return ErrInvalidIssuedAt()
```

**Reply by KevinJTaylor** [reply]
No effect i.e. same "iat NOT satisfied" issue

Not sure if this helps but if I print out the seconds and nonoseconds for now and the token issued at, 
```
			ns := now.Second()
			nns := now.Nanosecond()
			ts := sTtv.Second()
			tns := sTtv.Nanosecond()
```
I get 
- now: 48/982237000
- ttv: 49/0

**Reply by lestrrat** [reply]
Your test case doesn't run on its own, so I couldn't verify. On the other hand I tried running the tests in validate_test.go that explicitly test `iat` 1000 times each, and I can't reproduce :/

**Reply by KevinJTaylor** [reply]
What I don't get is how 'now' can be after 'ttv' tbh, but I guess that is pretty obvious ... in effect any rounding or truncation will not help if the two times are across the 'second barrier' (unless it was rounded/truncated to a minute, which doesn't make sense)

I'm inclined to look at the oauth library (idtoken) to see specifically what that does and see if that helps ... I do also have it at the back of my mind that although this may be the problem with my tests it may not be the occasional problem I get when running using the real service.

**Reply by KevinJTaylor** [reply]
I can give you the ks code (which is really just a provider for KeySetFrom), see below. The only other missing piece is the service account definition. I exported this from the google api console and then created the Mock service account from it (which I use in my tests)

```
type ks struct{}

func (k *ks) KeySetFrom(token jwt.Token) (jwk.Set, error) {
	discoveryDocument, err := getJSON("https://accounts.google.com/.well-known/openid-configuration")
	if err != nil {
		return nil, errors.Wrap(err, "cannot retrieve discovery document")
	}

	rawKeySet, err := getJSON(discoveryDocument[JWKsURI].(string))
	if err != nil {
		return nil, errors.Wrap(err, "cannot retrieve key set")
	}

	keySet := jwk.NewSet()
	for _, element := range rawKeySet["keys"].([]interface{}) {
		rawKey, err := json.Marshal(element)
		if err != nil {
			return nil, errors.Wrap(err, "cannot encode key")
		}

		key, err := jwk.ParseKey(rawKey)
		if err != nil {
			return nil, errors.Wrap(err, "cannot parse key")
		}

		keySet.Add(key)
	}
	return keySet, nil
}
```

**Reply by lestrrat** [reply]
Hmmm, well, I guess `idtoken.NewTokenSource()` makes a remote request eventually, so then my next guess may just be that your clocks are not exactly in sync, and that you may need to add some skew? `jwt.WithAcceptableSkew(5*time.Seconds)` maybe?

**Reply by KevinJTaylor** [reply]
Ahhh ok, that does make perfect sense
... thanks for your all your help @lestrrat, you've been jwx-cellent ;)
